nixos/users-groups: move home directory creation to systemd-tmpfiles

HeitorAugustoLN opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Fixes #6481
Based on the original work done in #223932 with some improvements

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

HeitorAugustoLN committed:
nixos/users-groups: move home directory creation to systemd-tmpfiles

The creation of user home directories was previously handled by an
activation script. This could lead to a race condition where home
directories were created in the /home mountpoint directory before the
/home filesystem was mounted.

This commit refactors this logic to use `systemd-tmpfiles` instead.
A `tmpfiles.d` rule is now generated for each user with `createHome = true`.
`systemd-tmpfiles` runs after filesystems are mounted, which resolves
the race condition.

This approach is based on the original work by John Soo in #223932

Co-authored-by: John Soo <john.soo@arista.com>

HeitorAugustoLN committed:
doc/release-notes: add release note for users-groups tmpfiles refactor

Adds a release note for the change that moved user home directory
creation to systemd-tmpfiles, resolving a race condition.

ElvishJerricco reviewed (changes requested):
I still think this is a good idea and would like to see it done. But since the original work, I believe both systemd-sysusers and userborn have arrived as alternatives to `update-users-groups.pl`. IIUC looking at other code, these tmpfiles need to be disabled when the other things are used.

/cc @nikstur

Granted, I'm not entirely convinced we shouldn't be using tmpfiles in those cases all the same.

EDIT: Actually it looks like both of those also already use tmpfiles themselves. Can we just unify all three of these then?

HeitorAugustoLN committed:
nixos/users-groups: unify home directory creation

The logic for creating user home directories was duplicated across the
users-groups, userborn, and sysusers modules.

This commit centralizes the logic into the users-groups module,
removing redundancy and improving maintainability. The implementation
now uses the structured `systemd.tmpfiles.settings` format.

nikstur reviewed (commented):
I can't say anything about the Perl code and I wonder whether it's worth it to change it at all. I want Userborn to eventually entirely replace the Perl script.

I encourage you to use `services.userborn.enable = true`. Then everything already works as expected.

Ma27 reviewed (commented):
Please retarget to staging-nixos since this causes a rebuild of all nixos tests.
Everything in there will usually be merged with the next kernel update, i.e. ~weekly.

See also https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#changes-rebuilding-all-nixos-tests
